tests/transport/test_cancel.cpp

```diff
@@ -0,0 +1,100 @@
+#include <vector>
+#include <iomanip>
+#include <utility>
+
+#include <gtest/gtest.h>
+
+#include <ghex/threads/none/primitives.hpp>
+#ifdef GHEX_TEST_USE_UCX
+#include <ghex/transport_layer/ucx/context.hpp>
+using transport = gridtools::ghex::tl::ucx_tag;
+#else
+#include <ghex/transport_layer/mpi/context.hpp>
+using transport = gridtools::ghex::tl::mpi_tag;
+#endif
+
+using threading = gridtools::ghex::threads::none::primitives;
+using context_type = gridtools::ghex::tl::context<transport, threading>;
+
+
+template<typename Message, typename Context>
+bool test_1(Context& context, unsigned int size) {
+    auto comm = context.get_communicator(context.get_token());
+    EXPECT_TRUE(comm.size() == 4);
+    Message msg(size*sizeof(int));
+
+    if (comm.rank() == 0) {
+        int* data = reinterpret_cast<int*>(msg.data());
+        for (unsigned int i = 0; i < size; ++i)
+            data[i] = i;
+        std::array<int, 3> dsts = {1,2,3};
+        auto futures = comm.send_multi(msg, dsts, 42+42);
+        for (auto& fut : futures)
+            fut.wait();
+        comm.barrier();
+        return true;
+    }
+    else {
+        auto fut = comm.recv(msg, 0, 42);
+        bool ok = fut.cancel();
+        comm.recv(msg, 0, 42+42).wait();
+        comm.barrier();
+        return ok;
+    }
+}
+
+TEST(cancel, future) {
+    auto context_ptr = gridtools::ghex::tl::context_factory<transport,threading>::create(1, MPI_COMM_WORLD);
+    auto& context = *context_ptr;
+
+    EXPECT_TRUE(test_1<std::vector<unsigned char>>(context,1));
+    EXPECT_TRUE(test_1<std::vector<unsigned char>>(context,32));
+    EXPECT_TRUE(test_1<std::vector<unsigned char>>(context,4096));
+}
+
+template<typename Message, typename Context>
+bool test_2(Context& context, unsigned int size) {
+    auto comm = context.get_communicator(context.get_token());
+    EXPECT_TRUE(comm.size() == 4);
+    Message msg(size*sizeof(int));
+
+    if (comm.rank() == 0) {
+        int* data = reinterpret_cast<int*>(msg.data());
+        for (unsigned int i = 0; i < size; ++i)
+            data[i] = i;
+        std::array<int, 3> dsts = {1,2,3};
+        auto futures = comm.send_multi(msg, dsts, 42+42);
+        for (auto& fut : futures)
+            fut.wait();
+        comm.barrier();
+        return true;
+    }
+    else {
+        using comm_t = typename Context::communicator_type;
+        using msg_t = typename comm_t::message_type;
+        using rank_t = typename comm_t::rank_type;
+        using tag_t = typename comm_t::tag_type;
+
+        int counter = 0;
+        auto req = comm.recv(msg, 0, 42, [&counter](msg_t, rank_t, tag_t){ ++counter; });
+        bool ok = req.cancel();
+        for (int i=0; i<100; ++i)
```

Comment:
Why `100`?

Reply:
a significant amount of times - 100 is an arbitrary value. I could add a timer instead, would that be better?

Reply:
What would be a reasonable amount of time? My problem here is: can this test fails just because the time is too short?

Reply:
It is more the other way round: you could miss a fail when not waiting long enough - a false positive.

Reply:
@boeschf Why do you need the loop here anyway? the canceled recv expects tag 42, which you don't send. So counter will always be 0, and although canceling of a UCX request requires a call to progress, we return true and don't really wait for it to happen.

Reply:
I guess we could return some more status information from `progress()` -  not a single number but number of cancels etc. what do you think? this way we could also eliminate the sometimes spurious results from ucx when we detect ucx-internal messages which we didn't send.

Reply:
I do not know if this can help, but returning something related to the internal state of the queues could be used there. Of course, if the queues are empty this does not mean that there are not messages still in flight somewhere

Reply:
@boeschf we don't have control over what the ucx worker progress returns, and we don't know what has finished during a progress call. I think it will be problematic to figure this out in general for shared workers. 

We also don't control whether the request is canceled or completed - what we know is that after a call to `cancel` it will eventually be either completed, or canceled, and a callback will be called with the corresponding state (in UCX). If you want to make sure that it has actually been canceled, then we could set some flag in the communicator recv callback.

Reply:
@angainor I believe by tracking the progress (number of called callbacks/number of cancels) ourselves we are safe. take a look at the suggested implementation in the new commits and let me know what you think.